### PR TITLE
s2u-36 5.2.1.1 Announcements: Ability to assign announcements to roles -

### PR DESCRIPTION
**Pull request by JuanDavid102**
fix

### CODE REVIEW

**Comment by mpellicer** [comment]
Does this fix cover bullhorns and also emails ?

**Comment by JuanDavid102** [comment]
About email, I dont know, but the bullhorns yes

**Comment by mpellicer** [comment]
@JuanDavid102 Check emails please, I think they are complaining about emails.

**Comment by JuanDavid102** [comment]
Ok

**Comment by mpellicer** [comment]
@JuanDavid102 Please review the build, is not building well.

**Comment by JuanDavid102** [comment]
Ok. I'm trying now with the complete deploy, but it looks like there is no problem, in my localhost

**Comment by jesusmmp** [comment]
> Ok. I'm trying now with the complete deploy, but it looks like there is no problem, in my localhost

Do a push --force again

**Comment by JuanDavid102** [comment]
It looks like fail in

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 import java.time.Instant;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Vector;
```
Comment (mpellicer) [inline announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:27]: Let's not use Vector, please, use list instead.

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java [hunk 2]
```diff
@@ -408,12 +409,57 @@ private String getAddress(Event event, AddressField field)
...
 	 * @param ref
 	 *        The entity reference.
+	 * @param site
+	 * 		The site whose membership the users will be deleted to
+	 */
+	protected void removeSpecialRecipients(List<User> users, Reference ref, Site site)
+	{
+		final AnnouncementMessage msg = (AnnouncementMessage) ref.getEntity();
+
+		for (Iterator<User> i = users.listIterator(); i.hasNext();) {
```
Comment (mpellicer) [inline announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:444]: Don't use iterators, iterate the list instead.

Where did you get this from?

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java [hunk 3]
```diff
@@ -408,12 +409,57 @@ private String getAddress(Event event, AddressField field)
...
+	protected List<User> getRecipients(Event event)
+	{
+		Reference ref = entityManager.newReference(event.getResource());
+		String siteId = (getSite() != null) ? getSite() : ref.getContext();
+		try {
+			Site site = siteService.getSite(siteId);
+			List<User> users = super.getRecipients(event);
+
+			// remove any other users
+			removeSpecialRecipients(users, ref, site);
```
Comment (mpellicer) [inline announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:423]: removeNotAllowedRecipients is a better name

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java (hunk 2 above)
Comment (mpellicer) [inline announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:444]: Let's use filters and Collectors to cleanup the list!

Comment (mpellicer) [reply announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:444]: Or maybe do it with a for loop setting a new list of allowed users, makes sense?

**Review by mpellicer — CHANGES_REQUESTED** [review]
Looks great @JuanDavid102, minor comments on the implementation.

**Comment by mpellicer** [comment]
Okay, the build looks broken because of #11440 , we'll review that in a different PR.

image

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java (hunk 2 above)
Comment (JuanDavid102) [reply announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:444]: I got it from the addSpecialRecipients of the parent class

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java [hunk 4]
```diff
@@ -408,12 +409,57 @@ private String getAddress(Event event, AddressField field)
...
+
+	/**
+	 * Remove in the user list any other users who shouldn't be notified
 	 * 
 	 * @param users
-	 *        The user list, already populated based on site visit and resource ability.
+	 * 		  a list of users for message
 	 * @param ref
 	 *        The entity reference.
+	 * @param site
```
[the remark this answers is not in the document]
Comment (JuanDavid102) [reply announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:437]: Ok

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java [hunk 5]
```diff
@@ -408,12 +409,51 @@ private String getAddress(Event event, AddressField field)
...
+	 * 		  The site where the announcement resides
+	 */
+	protected void removeNotAllowedRecipients(List<User> users, Reference ref, Site site)
+	{
+		final AnnouncementMessage msg = (AnnouncementMessage) ref.getEntity();
+		users.removeIf( u -> {
+			if (site.getMember(u.getId()) != null) {
+				for (String role : msg.getProperties().getPropertyList("selectedRoles")) {
```
Comment (mpellicer) [inline announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/SiteEmailNotificationAnnc.java:445]: This is not safe and will generate a NPE.

**Comment by mpellicer** [comment]
Please, rebase with the latest changes from #11460 to check that builds fine, thanks!

**Comment by mpellicer** [comment]
This is very nice!

**Pull request merged**